<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"
    />

    <title>UBC iGEM: Software Slides</title>

    <link rel="stylesheet" href="dist/reset.css" />
    <link rel="stylesheet" href="dist/reveal.css" />
    <link rel="stylesheet" href="dist/theme/igem.css" />
    <link rel="stylesheet" href="plugin/highlight/monokai.css" />
  </head>

  <body>
    <div class="reveal">
      <div class="slides">
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/bits_background.svg"
        >
          <div class="row">
            <div class="col">
              <h2>Enhancing DNA Storage with Synthetic Biology: Software</h2>
              <h3 class="highlight"><b>UBC iGEM</b></h3>
            </div>
          </div>
        </section>

        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Project Description</h2>
          <p class="r-fit-text">
            Our project aims to tackle the growing need for a better, more
            energy-efficient data storage medium compared to current magnetic
            and optical data storage options by means of synthetic biology.
            Currently, we aim to achieve this through 2 separate tracks:
          </p>
          <p class="r-fit-text fragment">
            Developing an <b>enzymatic </b>DNA synthesis platform that can
            elongate a <b>single-stranded </b> DNA (ssDNA) in a
            <b>template-independent </b> manner. The synthesized ssDNA strand
            will then be converted to a more stable, double-stranded DNA (dsDNA)
            and inserted into a plasmid for long-term data storage.
          </p>
          <p class="r-fit-text fragment">
            Developing a <b>data encoding/decoding pipeline</b> that allows
            binary files (used by computers) to be stored in a ternary format
            compatible with our DNA synthesis platform, retrieved, and converted
            back into binary.
          </p>
        </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Goals</h2>
          <div class="r-fit-text fragment">
            <h3>In Silico:</h3>
            Demonstrate ability to encode and decode information someone may
            store in long-term storage, in the 1000s of nucleotides long.
          </div>
          <div class="r-fit-text fragment">
            <h3>With wet lab:</h3>
            Demonstrate ability to encode and decode a 100 nucleotide sequence
            with 30% error.
          </div>
        </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Plan</h2>
          <div class="fragment r-fit-text">
            <h3>DBTL 1: March to April</h3>
            <p>
              Implement a barebones pipeline, and see how much error can be
              tolerated in 100 nucleotide long DNA sequences with in silico
              testing.
            </p>
          </div>
          <div class="fragment r-fit-text">
            <h3>DBTL 2: April to May</h3>
            <p>
              Redefine algorithms to tolerate up to 30% error in 100 nucleotide
              long DNA sequences, with in silico testing.
            </p>
          </div>
                 </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/molecules_background.svg"
        >
          <h2>DBTL 1: Proof of Concept</h2>
          <ul>
            <li>
            Encoding
          </li>
      <li>
            Error Correction
          </li>
      <li>
            Decoding
          </li>
      <li>
            ChaosDNA
          </li>

          </ul>        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/encoding.svg"
        >
          <h2>Encoding</h2>
          <ul>
            <li>Primer Generation</li>
            <li>Sequence Generation</li>
          </ul>
        </section>

        <section
  data-auto-animate
          data-background="public/backgrounds/encoding.svg"

        ><h2>Primer Generation</h2></section>

        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg">
          <h2>Why generate our own primers?</h2>
          <ul>
            <li>act as unique identifiers for information</li>
            <li>required for PCR amplification</li>
            <li>specify requirements for TdT enzyme</li>
          </ul>
          <p>How? Using a genetic algorithm.</p>
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/info_background.svg">
          <h2>Why use a genetic algorithm?</h2>
          <ul>
            <li>type of optimization algorithm</li>
            <li>
              uses a set of constraints to produce heuristics to determine best
              parent candidates, which go on to produce children population
            </li>
          </ul>
          <p>Primer design fits this description!</p>
        </section>


        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg">
          <h2>Schematic</h2>
          <img class="stretch" src="public/images/schematic-primer-gen.png" />
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg"
        >
          <h2>Example: two chosen parents</h2>
          <img src="public/images/primer_gen.png" />
        </section>


        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg"
        >
          <h2>Sequence Generation</h2>
        </section>

        <section
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Binary to Nucleotides</h2>
          <p>Some limitations...</p>
          <ul>
            <li>short strands (100 nt)</li>
            <li>semi-specific synthesis</li>
            <li>high rate of deletion errors (30%)</li>
          </ul>
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Short strands?</h2>
          <p>Blocked data and conversion to ternary</p>
          <img class="stretch" src="public/images/encode_1.png" />
          <p>Part 1: proposed encoding scheme</p>
        </section>

        <section

          data-auto-animate
 data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2 class="r-fit-text">Short strands?</h2>
          <p>Text Compression</p>
          <img class="stretch" src="public/images/ts_zip-compression.png" />
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2>Short strands?</h2>
          <p>Vector Data Formats</p>
          TODO
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2>Semi-specific synthesis?</h2>
          <p>Rotation based cipher</p>
          <img class="stretch" src="public/images/rbc.png" />
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2 class="r-fit-text">Rotation based cipher: One char</h2>
          <img class="stretch" src="public/images/one_char.png" />
        </section>

        <section

          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2>High rate of deletion errors?</h2>
          <p class="r-fit-text">For the first iteration, we want to see what percentage of deletion errors we can correct for with minimal error correction. Some ways of reducing rate of deletion errors or preventing deletion errors include:</p>
          <ul class="r-fit-text">
            <li><b>synthesizing shorter sequences (more blocks of shorter length)</b></li>
            <li>more complex encoding methods</li>
            <li>more complex error correction methods (inner and outer codes)</li>
            <li><b>choose the most stable DNA strand (starting nucleotide) based on thermodynamic properties</b></li>
            <li><b>collect metadata and generate a checksum, to be used to sysematically resolve deletion errors</b></li>
        </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/molecules_background.svg">
          <h2>Part 2: proposed encoding scheme</h2>
          <img class="stretch" src="public/images/encode_2.png"/>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2>Entire encoding pipeline</h2>
          <img src="public/images/seq_gen.png" />
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/bits_background.svg">
          <h2>Decoding</h2>
          <ul>
            <li>Sequence Alignment</li>
            <li>Error Correction</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/bits_background.svg">
          <h2>Sequence Alignment</h2>
          <ul><li>Sanger Sequencing</li><li><b>NGS</b></li></ul></section>

             <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg">
          <h2>De Novo Assembly: NGS</h2>
          <img class="stretch" src="public/images/seq_align.png">
                      </section>


         <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg">
          <h2>ChaosDNA</h2>
        </section>

        <section data-auto-animate>
          <h2>DBTL 3 and further</h2>
 <div class="fragment r-fit-text">
            <h3>DBTL 3: May to June</h3>
            <p>
              Implement DNA Storage Alliance specifications, and do in silico
              testing on DNA sequences with 1000s of nucleotides.
            </p>
          </div>
          <div class="fragment r-fit-text">
            <h3>DBTL 4 and 5: June to July</h3>
            <p>
              Test our software on sequences synthesized by wet lab, and
              redefine algorithms with in silico testing and wet lab data.
            </p>
          </div>

        </section>
      </div>
    </div>

    <script src="dist/reveal.js"></script>
    <script src="plugin/notes/notes.js"></script>
    <script src="plugin/markdown/markdown.js"></script>
    <script src="plugin/highlight/highlight.js"></script>
    <script>
      // More info about initialization & config:
      // - https://revealjs.com/initialization/
      // - https://revealjs.com/config/
      Reveal.initialize({
        hash: true,

        // Learn about plugins: https://revealjs.com/plugins/
        plugins: [RevealMarkdown, RevealHighlight, RevealNotes],
      });
    </script>
  </body>
</html>
