<html lang="en">
  <head>
    <meta charset="utf-8" />
    <meta
      name="viewport"
      content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no"
    />

    <title>UBC iGEM: Software Slides</title>

    <link rel="stylesheet" href="dist/reset.css" />
    <link rel="stylesheet" href="dist/reveal.css" />
    <link rel="stylesheet" href="dist/theme/igem.css" />
    <link rel="stylesheet" href="plugin/highlight/monokai.css" />
  </head>

  <body>
    <div class="reveal">
      <div class="slides">
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/bits_background.svg"
        >
          <div class="row">
            <div class="col">
              <h2>Enhancing DNA Storage with Synthetic Biology: Software</h2>
              <h3 class="highlight"><b>UBC iGEM</b></h3>
            </div>
          </div>
        </section>

        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Project Description</h2>
          <p class="r-fit-text">
            Our project aims to tackle the growing need for a better, more
            energy-efficient data storage medium compared to current magnetic
            and optical data storage options by means of synthetic biology.
            Currently, we aim to achieve this through 2 separate tracks:
          </p>
          <p class="r-fit-text fragment">
            Developing an <b>enzymatic </b>DNA synthesis platform that can
            elongate a <b>single-stranded </b> DNA (ssDNA) in a
            <b>template-independent </b> manner. The synthesized ssDNA strand
            will then be converted to a more stable, double-stranded DNA (dsDNA)
            and inserted into a plasmid for long-term data storage.
          </p>
          <p class="r-fit-text fragment">
            Developing a <b>data encoding/decoding pipeline</b> that allows
            binary files (used by computers) to be stored in a ternary format
            compatible with our DNA synthesis platform, retrieved, and converted
            back into binary.
          </p>
        </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Goals</h2>
          <div class="r-fit-text fragment">
            <h3>In Silico:</h3>
            Demonstrate ability to encode and decode information someone may
            store in long-term storage, in the 1000s of nucleotides long.
          </div>
          <div class="r-fit-text fragment">
            <h3>With wet lab:</h3>
            Demonstrate ability to encode and decode a 100 nucleotide sequence
            with 30% error.
          </div>
        </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Plan</h2>
          <div class="fragment r-fit-text">
            <h3>DBTL 1: March to April</h3>
            <p>
              Implement a barebones pipeline, and see how much error can be
              tolerated in 100 nucleotide long DNA sequences with in silico
              testing.
            </p>
          </div>
          <div class="fragment r-fit-text">
            <h3>DBTL 2: April to May</h3>
            <p>
              Redefine algorithms to tolerate up to 30% error in 100 nucleotide
              long DNA sequences, with in silico testing.
            </p>
          </div>
        </section>
        <section
          data-auto-animate
          class="left-align"
          data-background="public/backgrounds/molecules_background.svg"
        >
          <h2>DBTL 1: Proof of Concept</h2>
          <ul class="fragment">
            <li>Encoding</li>
            <li>Error Correction</li>
            <li>Decoding</li>
            <li>ChaosDNA</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/crab_background.svg"
        >
          <h2>What programming language are we using?</h2>
          <h3 class="fragment">Rust</h3>
          <ul>
            <li class="fragment">
              systems programing language that is fast, memory efficient and
              memory-safe
            </li>
            <li class="fragment">
              expressive type system, great documentation, robust tooling
            </li>
            <li class="fragment">most admired language among developers</li>
          </ul>
          <cite
            ><a
              href="https://github.blog/2023-08-30-why-rust-is-the-most-admired-language-among-developers/"
              >https://github.blog/2023-08-30-why-rust-is-the-most-admired-language-among-developers/</a
            ></cite
          >
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/encoding.svg"
        >
          <h2>Encoding</h2>
          <ul>
            <li>Primer Generation</li>
            <li>Sequence Generation</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/encoding.svg"
        >
          <h2>Primer Generation</h2>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg"
        >
          <h2>Why generate our own primers?</h2>
          <ul>
            <li>act as unique identifiers for information</li>
            <li>required for PCR amplification</li>
            <li>specify requirements for TdT enzyme</li>
          </ul>
          <p class="fragment">How? Using a genetic algorithm.</p>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Why use a genetic algorithm?</h2>
          <ul>
            <li>type of optimization algorithm</li>
            <li>
              uses a set of constraints to produce heuristics to determine best
              parent candidates, which go on to produce children population
            </li>
          </ul>
          <p class="fragment">Primer design fits this description!</p>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Schematic</h2>
          <img class="stretch" src="public/images/schematic-primer-gen.png" />
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg"
        >
          <h2>Example: two chosen parents</h2>
          <img src="public/images/primer_gen.png" />
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/default_background.svg"
        >
          <h2>Sequence Generation</h2>
        </section>

        <section
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Binary to Nucleotides</h2>
          <p>Many encoding formats exist:</p>
          <ul>
            <li class="fragment">
              Base 4 Encoding: 0 → A, 1 → T, 2 → G, 3 → C
            </li>
            <li class="fragment">Church Encoding: 0 → A or C, 1 → T or G</li>
            <li class="fragment">
              Base 2 Encoding: 00 → A, 11 → T, 01 → G, 10 → C
            </li>
            <li class="fragment">
              HEDGES (key-autokey cipher) ECC: (hash(input) + bit)mod4 → Base 4
              Encoding
            </li>
          </ul>
          <p class="fragment">
            These encoding methods will ultimately be tested in silico...
          </p>
        </section>

        <section
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Binary to Nucleotides</h2>
          <p>Some limitations...</p>
          <ul>
            <li class="fragment">short strands (100 nt)</li>
            <li class="fragment">high rate of deletion errors (30%)</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Short strands?</h2>
          <p>Blocking data</p>
          <img class="stretch" src="public/images/block.png" />
          <ul class="r-fit-text fragment">
            <li class="fragment">
              requires we generate unique primers per strand
            </li>
            <li class="fragment">
              allows for parallel synthesis of DNA strands
            </li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Short strands?</h2>
          <p class="r-fit-text fragment">
            To encode the UBC iGEM sponsorship package...
          </p>
          <div class="row fragment">
            <div style="text-wrap: wrap" class="left-align">
              <ul>
                <li>
                  1.8 MB → 1.8 10<sup>6</sup> bytes → 1.44 * 10<sup>7</sup> bits
                  → 180000 strands of DNA
                </li>
                <li>
                  assuming each DNA strand's information portion is 80 unique
                  bases and we are using HEDGES Encoding
                </li>
                <li class="fragment">we will require compression...</li>
              </ul>
            </div>
            <img class="side-img stretch" src="public/images/example.png" />
          </div>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Short strands?</h2>
          <img class="stretch" src="public/images/ts_zip-compression.png" />
        </section>

        <!-- <section
             data-auto-animate
             data-background="public/backgrounds/dna_right_corner_backgroud.svg"
             >
             <h2>Semi-specific synthesis?</h2>
             <p>Rotation based cipher</p>
             <img class="stretch" src="public/images/rbc.png" />
             </section>
        -->

        <!-- <section
             data-auto-animate
             data-background="public/backgrounds/dna_right_corner_backgroud.svg"
             >
             <h2 class="r-fit-text">Rotation based cipher: One char</h2>
             <img class="stretch" src="public/images/one_char.png" />
             </section>
        -->

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>High rate of deletion errors?</h2>
          <p class="r-fit-text left-align">
            For the first iteration, we want to see what percentage of deletion
            errors we can correct for with minimal error correction. Some ways
            of reducing rate of deletion errors or preventing deletion errors
            include that we will explore in further DBTLs are:
          </p>
          <ul class="r-fit-text">
            <li class="fragment">
              <b
                >synthesizing shorter sequences (more blocks of shorter
                length)</b
              >
            </li>
            <li class="fragment">more complex encoding strategies</li>
            <li class="fragment">
              <b>
                more complex error correction methods (inner and outer codes)
              </b>
            </li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>High rate of deletion errors?</h2>
          <p>HEDGES error-correcting codes: Why?</p>
          <ul>
            <li>
              corrects for mutations and deletions, not prevalent type of error
              in traditional media storage
            </li>
            <li>form of inner code: bits that encode for information</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>HEDGES error-correcting codes</h2>
          <img class="stretch" src="public/images/hedges-encode.jpg" />
          <p>Redundacy is enforced in the encoding strategy.</p>
          <cite class="citation">
            Press, W. H., Hawkins, J. A., Jones, S. K., Schaub, J. M., &amp;
            Finkelstein, I. J. (2020). HEDGES error-correcting code for DNA
            storage corrects indels and allows sequence constraints.
            <em>Proceedings of the National Academy of Sciences</em>,
            <em>117</em>(31), 18489–18496.
            <a href="https://doi.org/10.1073/pnas.2004821117"
              >https://doi.org/10.1073/pnas.2004821117</a
            >
          </cite>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>What does this mean?</h2>
          <p>
            HEDGES uses a hash function to encode redundancy and generate the
            base to be synthesized.
          </p>
          <p class="fragment"><b>What's a hash function?</b></p>
          <ul>
            <li class="fragment">
              maps data of arbitrary size to a fixed-size value
            </li>
            <li class="fragment">fast to compute hash values</li>
            <li class="fragment">
              very low chance of returning the same hash value for two different
              hash inputs
            </li>
          </ul>
          <p>
            We will also use a established checksum algorithm to generate a
            checksum to signal if error correction is needed.
          </p>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>HEDGES error-correcting codes: Rationale</h2>
          <p>
            "Hashing each bit value withits strand ID, bit index, and a few
            previous bits “poisons” bad decoding hypotheses, allowing for
            correction of indels."
          </p>
          <p>
            "In summary, the algorithm encodes information as a stream of
            nucleotides such that any single decoding error in either nucleotide
            identity or nucleotide position will “poison” the downstream
            predictions. Thus, on decoding, there will be onlyone good-scoring
            chain of guesses—the correct one."
          </p>
          <cite class="citation">
            Press, W. H., Hawkins, J. A., Jones, S. K., Schaub, J. M., &amp;
            Finkelstein, I. J. (2020). HEDGES error-correcting code for DNA
            storage corrects indels and allows sequence constraints.
            <em>Proceedings of the National Academy of Sciences</em>,
            <em>117</em>(31), 18489–18496.
            <a href="https://doi.org/10.1073/pnas.2004821117"
              >https://doi.org/10.1073/pnas.2004821117</a
            >
          </cite>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>HEDGES error-correcting codes: Rationale</h2>
          <ul>
            <li>
              designed with orthogonality in mind: can add other error
              correction codes without interference, such as RSC
            </li>
            <li>
              variable paramters, so we can tune to our specific DNA synthesis
            </li>
            <li>MIT license, implementions online in C++ and Python</li>
          </ul>
          <br />
          <br />
          <cite class="citation">
            Press, W. H., Hawkins, J. A., Jones, S. K., Schaub, J. M., &amp;
            Finkelstein, I. J. (2020). HEDGES error-correcting code for DNA
            storage corrects indels and allows sequence constraints.
            <em>Proceedings of the National Academy of Sciences</em>,
            <em>117</em>(31), 18489–18496.
            <a href="https://doi.org/10.1073/pnas.2004821117"
              >https://doi.org/10.1073/pnas.2004821117</a
            >
          </cite>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/bits_background.svg"
        >
          <h2>Decoding</h2>
          <ul>
            <li>Sequence Alignment</li>
            <li>Error Correction</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/bits_background.svg"
        >
          <h2>Sequence Alignment</h2>
          <p>we will most likely use either (or combination of):</p>
          <ul>
            <li class="fragment">Sanger Sequencing</li>
            <li class="fragment"><b>NGS</b></li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>De Novo Assembly</h2>
          <p>Why implement de novo assembly?</p>

          <ul>
            <li>no reference template, so de novo assembly is required</li>
            <li>
              important proof of concept to demonstrate our software can put
              back together a DNA sequence of 1000s of bases long
            </li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>De Novo Assembly</h2>
          <img class="stretch" src="public/images/seq_align.png" />
          <p>greedy graph search</p>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Why greedy graph search?</h2>
          <ul>
            <li class="fragment">the exact solution is NP-hard problem</li>
            <li class="fragment">
              our sequences will be at most 1000-3000 bases long, so being
              greedy will usually yield the exact solution anyways
            </li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Error Correction</h2>
          <p>How does this work?</p>
          <ul>
            <li>
              we are given a DNA sequence, and assuming we have correctly
              decoded bits[0...i-1], and we want bits[i]
            </li>
            <li>
              we use the hash function from encoding stage to "guess" what the
              correct base should be
            </li>
            <li>
              if we correctly guess the base, we continue searching that branch,
              otherwise assign a cumulative penality score or abandon that
              branch
            </li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>HEDGES error-correction code: Reminder</h2>
          <img class="stretch" src="public/images/full-hedges.png" />
          <br />
          <cite class="citation">
            Press, W. H., Hawkins, J. A., Jones, S. K., Schaub, J. M., &amp;
            Finkelstein, I. J. (2020). HEDGES error-correcting code for DNA
            storage corrects indels and allows sequence constraints.
            <em>Proceedings of the National Academy of Sciences</em>,
            <em>117</em>(31), 18489–18496.
            <a href="https://doi.org/10.1073/pnas.2004821117"
              >https://doi.org/10.1073/pnas.2004821117</a
            >
          </cite>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Error Correction</h2>
          <img class="stretch" src="public/images/hedges-decode.jpg" />
          <cite class="citation">
            Press, W. H., Hawkins, J. A., Jones, S. K., Schaub, J. M., &amp;
            Finkelstein, I. J. (2020). HEDGES error-correcting code for DNA
            storage corrects indels and allows sequence constraints.
            <em>Proceedings of the National Academy of Sciences</em>,
            <em>117</em>(31), 18489–18496.
            <a href="https://doi.org/10.1073/pnas.2004821117"
              >https://doi.org/10.1073/pnas.2004821117</a
            >
          </cite>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>ChaosDNA</h2>
          <ul>
            <li>an in-silico tool for generating faulty DNA sequences</li>
          </ul>
          <p>Why?</p>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>ChaosDNA</h2>
          <ul>
            <li>an in-silico tool for generating faulty DNA sequences</li>
            <li>
              will allow us to prove our software tool can deal with files of
              more realistic sizes (MB)
            </li>
            <li>
              and enable us to work with DNA sequences of varying levels of
              deletion, insertion and mutation errors
            </li>
          </ul>
          <img src="public/images/chaosdna.png" />
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>After DBTL 2?</h2>
          <div class="fragment r-fit-text">
            <h3>DBTL 3: May to June</h3>
            <p>
              Implement DNA Storage Alliance specifications, and do in silico
              testing on DNA sequences with 1000s of nucleotides.
            </p>
          </div>
          <div class="fragment r-fit-text">
            <h3>DBTL 4 and 5: June to July</h3>
            <p>
              Test our software on sequences synthesized by wet lab, and
              redefine algorithms with in silico testing and wet lab data.
            </p>
          </div>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/dna_right_corner_backgroud.svg"
        >
          <h2>Future Directions</h2>
          <ul>
            <li>DNA Storage Alliance specifications</li>
            <li>Outer Codes: GC</li>
            <li>Graphical User Interface</li>
            <li>SVGs</li>
          </ul>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>DNA Storage Alliance specifications</h2>
          <p>
            "Unlike traditional storage media such as tape, HDD, and SSD, DNA
            does not have a fixed physical structure, a built-in controller, or
            a way to address different regions of the media linearly, and thus
            needs a mechanism to start reading or “booting up” a DNA archive
            that does not rely on such a structure. The SNIA DNA Archive Rosetta
            Stone (DARS) working group, one of four working groups in the DNA
            Data Storage Alliance aimed at defining standards for DNA data
            storage systems, has developed two specifications to enable archive
            readers to find the sequence to begin booting up the data."
          </p>
          <cite
            ><a
              href="https://www.snia.org/news_events/newsroom/dna-data-storage-alliance-releases-its-first-specifications"
              >https://www.snia.org/news_events/newsroom/dna-data-storage-alliance-releases-its-first-specifications</a
            ></cite
          >
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Outer Codes: GuessCheck</h2>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>Graphical User Interface</h2>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/info_background.svg"
        >
          <h2>SVGs and QR Codes</h2>
          <div class="fragment r-fit-text">
            <h3>Scalable Vector Graphics</h3>
            <p>
              Store shapes as mathematical equations rather than individual pixels.
	      Further compressible through traditional mechanisms or
	      aforementioned text compression mechanisms.
            </p>
          </div>
          <div class="fragment r-fit-text">
            <h3>QR Codes</h3>
            <p>
              Designed to store redundant information, enabling extreme
	      error correction. Can be efficiently stored in many formats,
	      including SVG or PNG.
            </p>
          </div>
        </section>

        <section
          data-auto-animate
          data-background="public/backgrounds/bits_background.svg"
        >
          <h1>Thank you!</h1>
          <h3>Questions?</h3>
        </section>
      </div>
    </div>

    <script src="dist/reveal.js"></script>
    <script src="plugin/notes/notes.js"></script>
    <script src="plugin/markdown/markdown.js"></script>
    <script src="plugin/highlight/highlight.js"></script>
    <script>
      // More info about initialization & config:
      // - https://revealjs.com/initialization/
      // - https://revealjs.com/config/
      Reveal.initialize({
        hash: true,

        // Learn about plugins: https://revealjs.com/plugins/
        plugins: [RevealMarkdown, RevealHighlight, RevealNotes],
      });
    </script>
  </body>
</html>
